File: ruby/difference-of-squares/difference_of_squares.rb
module BookKeeping
  VERSION = 3
end

class Squares

  def initialize(number)
    @range = (0..number)
  end

  def square_of_sum
    @range.reduce(:+) ** 2
  end

  def sum_of_squares
    @range.reduce {|sum, n| sum + (n ** 2) }
  end

  def difference
    square_of_sum - sum_of_squares
  end

end

File: go/difference-of-squares/difference_of_squares.go
package diffsquares

// Returns the square of the sum of the first num natural number
func SquareOfSum(num int) int {
	sum := 0
	for i := num; i > 0; i-- {
		sum += i
	}
	return sum * sum
}

// Returns the sum of the squares of the first num natural numbers
func SumOfSquares(num int) int {
	sum := 0
	for i := num; i > 0; i-- {
		sum += i * i
	}
	return sum
}

// Returns the difference between the SquareOfSum and SumOfSquares
func Difference(num int) int {
	return SquareOfSum(num) - SumOfSquares(num)
}
